cfme/utils/template/base.py

```diff
@@ -0,0 +1,245 @@
+        self._provider_data = kwargs.get('provider_data')
+
+        self._cmd_line_args = cmd_line_args
+
+        self.stream = stream
+        self.provider = provider
+        self.template_name = template_name
+
+        self.kwargs = kwargs
+
+    @property
+    def stream_url(self):
+        """ Returns URL to image directory of a stream.
+
+        Example output:
+            http://hostname.redhat.com/builds/cfme/version/stable/
+
+        Default value:
+            stored in cfme_data.basic_info.cfme_images_url[stream] for known streams
+        """
+        urls = cfme_data.basic_info.cfme_images_url
+
+        if self._stream_url:
+            return self._stream_url
+        elif self.stream in urls:
+            return urls[self.stream]
+        else:
+            logger.error("Stream %s is not listed in cfme_data. "
+                         "Please specify stream URL with --stream_url", self.stream)
+            raise TemplateUploadException("Cannot get stream URL.")
+
+    @property
+    def image_url(self):
+        """ Returns URL to exact image file.
+
+        Example output:
+            http://hostname.redhat.com/builds/cfme/version/stable/cfme-type-version.x86_64.vhd
+
+        Default value:
+            browses stream_url for images and matches them to image_pattern regex
+        """
+        if not self._image_url:
+            try:
+                with closing(urlopen(self.stream_url)) as stream_dir:
+                    string_from_url = stream_dir.read()
+            except URLError as e:
+                logger.error("Cannot get image URL from %s: %s", self.stream_url, e.reason.strerror)
+                raise TemplateUploadException("Cannot get image URL.")
+            else:
+                image_name = self.image_pattern.findall(string_from_url)
+                if len(image_name):
+                    return "{}{}".format(self.stream_url, image_name[0])
+        return self._image_url
+
+    @property
+    def image_name(self):
+        """ Returns filename of an image.
+
+        Example output: cfme-type-version.x86_64.vhd
+        """
+        return self.image_url.split("/")[-1]
+
+    @property
+    def mgmt(self):
+        """ Returns wrapanapi management system class.
+
+        provider_data has higher priority than provider name.
+        """
+        if self.provider_data:
+            return get_mgmt(self.provider_data)
+        elif self.provider:
+            return get_mgmt(self.provider)
+        else:
+            logger.error("Please specify provider or provider_data to retrieve it's mgmt.")
+            raise TemplateUploadException("Cannot get_mgmt due to empty data.")
+
+    @property
+    def provider_data(self):
+        """ Returns AttrDict from cfme_data[management_systems][provider]."""
+        if not self._provider_data:
+            return cfme_data.management_systems[self.provider]
+        return self._provider_data
+
+    @property
+    def template_upload_data(self):
```

Comment:
Not a fan of naming here, since there is a yaml configuration block for template_upload, and this is a BaseTemplateUpload class.

I think my confusion ultimately stems from the class name though, given the class's modeling.

What do you think about calling this class `ProviderTemplateUpload` instead of 'Base', since its in effect modeling a trackerbot provider-template record.